### PR TITLE
br: restore tiflash replica count after PiTR

### LINKED ISSUE #37208: br: support restore tiflash replicas
## Enhancement
As title.

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close  #37208

Problem Summary:
Currently, we cannot restore the information of TiFlash replicas.

### What is changed and how it works?
This PR added some hooks to the PiTR restore process, and would collect the tilfash-related infromation and then generate and execute those DDLs for restoring the TiFlash replicas.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
The script: https://gist.github.com/YuJuncen/4f82d1769a766faa78098923d3e32071
Checksum matches:
```
[["310000", "154946.6817939191110397"], ["310000", "154947.6057966217137621"], ["310000", "154820.4573007339340533"]] 
vs 
[["310000", "154946.6817939191110397"], ["310000", "154947.6057966217137621"], ["310000", "154820.4573007339340533"]]; 
equals? = true
```
TiFlash replica matches:
```
irb(main):005:0> (SQL.new :downstream).sql 'select * from information_schema.tiflash_replica'
=> [["test", "table4", "215", "1", "", "1", "1", "NORMAL"], ["test", "table3", "216", "1", "", "1", "1", "NORMAL"]]
irb(main):006:0> SQL.new.sql 'select * from information_schema.tiflash_replica'
=> [["test", "table4", "193", "1", "", "1", "1", "NORMAL"], ["test", "table3", "196", "1", "", "1", "1", "NORMAL"]]
```
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
PiTR (Log Backup) can restore TiFlash replica metadata now.
```

### CODE REVIEW

File: br/pkg/stream/rewrite_meta_rawkv.go [hunk 1]
```diff
@@ -379,36 +393,74 @@ func (sr *SchemasReplace) rewriteEntryForAutoRandomTableIDKey(e *kv.Entry, cf st
...
+}
+
+// rewriteValueV2 likes rewriteValueV1, but provides a richer return value.
+func (sr *SchemasReplace) rewriteValueV2(value []byte, cf string, rewrite func([]byte) ([]byte, bool, error)) (rewriteResult, error) {
 	switch cf {
 	case DefaultCF:
-		return cbRewrite(value)
+		newValue, needRewrite, err := rewrite(value)
+		if err != nil {
+			return rewriteResult{}, nil
```
Comment (joccau) [inline br/pkg/stream/rewrite_meta_rawkv.go:408]:
```suggestion
			return rewriteResult{}, errors.Trace(err)
```

File: br/pkg/restore/tiflashrec/tiflash_recorder.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+func (r *TiFlashRecorder) Iterate(f func(tableID int64, replica model.TiFlashReplicaInfo)) {
+	for k, v := range r.items {
+		f(k, v)
+	}
+}
+
+func (r *TiFlashRecorder) Rewrite(oldID int64, newID int64) {
+	old, ok := r.items[oldID]
```
Comment (joccau) [inline br/pkg/restore/tiflashrec/tiflash_recorder.go:71]:
```suggestion
    if oldID == newID {
        return
    }
	old, ok := r.items[oldID]
```

File: br/pkg/task/stream.go [hunk 3]
```diff
@@ -1125,6 +1128,17 @@ func restoreStream(
 	if err != nil {
 		return errors.Trace(err)
 	}
+	schemasReplace.AfterTableRewritten = func(deleted bool, tableInfo *model.TableInfo) {
```
Comment (joccau) [inline br/pkg/task/stream.go:1131]: Can we put the function `AfterTableRewritten()` into client.InitSchemasReplaceForDDL() ?

**Review by joccau — COMMENTED** [review]
rest LGTM.

File: br/pkg/task/stream.go (hunk 3 above)
Comment (YuJuncen) [reply br/pkg/task/stream.go:1131]: Yep, however we must pass the `tiflashRecorder` into `InitSchemasReplaceForDDL`. So I think leave it outside would be fine?🤔

Comment (joccau) [reply br/pkg/task/stream.go:1131]: OK

**Review by joccau — APPROVED** [review]
LGTM

**Review by Leavrth — APPROVED** [review]

**Pull request merged**